File: scripts/migrate_users.py
#!/usr/bin/env python3
"""
用户数据迁移脚本 (Python版)

将现有的环境变量用户迁移到 SQLite 数据库
默认密码: root

使用方法:
    python3 scripts/migrate_users.py [db_path]
    
示例:
    python3 scripts/migrate_users.py                    # 使用默认路径 data/jobs.db
    python3 scripts/migrate_users.py /path/to/custom.db # 指定数据库路径
"""

import sys
import os
import hashlib
import secrets
import sqlite3
from datetime import datetime
from pathlib import Path


def generate_salt(length: int = 32) -> str:
    """生成随机盐值"""
    return secrets.token_hex(length)


def hash_password(password: str, salt: str) -> str:
    """SHA256 + Salt 哈希"""
    salted = f"{password}{salt}".encode('utf-8')
    return hashlib.sha256(salted).hexdigest()


# 默认密码 (用户指定)
DEFAULT_PASSWORD = "root"

# 要迁移的用户列表
USERS_TO_MIGRATE = [
    {
        'username': 'admin',
        'role': 'admin',
        'permissions': [
            'view_jobs', 'view_stats', 'manage_crawler', 'view_system',
            'use_recommendations', 'manage_users',
            'jobs:read', 'jobs:write', 'crawler:read', 'crawler:write',
            'system:read', 'system:write', 'users:read', 'users:write',
            'recommendations:read', 'recommendations:write',
            'match:read', 'match:write',
        ]
    },
    {
        'username': 'operator',
        'role': 'operator',
        'permissions': [
            'view_jobs', 'view_stats', 'manage_crawler', 'view_system',
            'use_recommendations',
            'jobs:read', 'jobs:write', 'crawler:read', 'crawler:write',
            'system:read', 'recommendations:read', 'recommendations:write',
            'match:read', 'match:write',
        ]
    },
    {
        'username': 'viewer',
        'role': 'viewer',
        'permissions': [
            'view_jobs', 'view_stats', 'view_system', 'use_recommendations',
            'jobs:read', 'crawler:read', 'system:read',
            'recommendations:read', 'match:read',
        ]
    }
]


def migrate_users(db_path: str):
    """执行用户迁移"""
    
    print("=" * 60)
    print("🚀 开始迁移用户到数据库...")
    print(f"📁 数据库路径: {db_path}")
    print(f"🔑 默认密码: {DEFAULT_PASSWORD}")
    print()
    
    db_path = Path(db_path)
    db_path.parent.mkdir(parents=True, exist_ok=True)
    
    try:
        conn = sqlite3.connect(str(db_path))
        cursor = conn.cursor()
        
        # 启用 WAL 模式
        cursor.execute("PRAGMA journal_mode = WAL")
        
        print("✅ 数据库连接成功\n")
        
        # 创建认证表
        cursor.executescript("""
            CREATE TABLE IF NOT EXISTS users (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                username TEXT UNIQUE NOT NULL,
                password_hash TEXT NOT NULL,
                salt TEXT NOT NULL,
                role TEXT NOT NULL DEFAULT 'viewer',
                permissions TEXT DEFAULT '[]',
                is_active INTEGER DEFAULT 1,
                failed_login_count INTEGER DEFAULT 0,
                locked_until TIMESTAMP NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                last_login_at TIMESTAMP NULL
            );
            
            CREATE TABLE IF NOT EXISTS sessions (
                id TEXT PRIMARY KEY,
                user_id INTEGER NOT NULL,
                token TEXT UNIQUE NOT NULL,
                ip_address TEXT,
                user_agent TEXT,
                expires_at TIMESTAMP NOT NULL,
                created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
                FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
            );
            
            CREATE INDEX IF NOT EXISTS idx_sessions_user_id ON sessions(user_id);
            CREATE INDEX IF NOT EXISTS idx_sessions_token ON sessions(token);
            CREATE INDEX IF NOT EXISTS idx_users_username ON users(username);
        """)
        
        print("✅ 认证表已就绪\n")
        
        # 检查现有用户
        cursor.execute("SELECT COUNT(*) FROM users")
        existing_count = cursor.fetchone()[0]
        
        if existing_count > 0:
            print(f"⚠️ 数据库中已有 {existing_count} 个用户")
            
            answer = input("是否继续覆盖？(y/N): ").strip().lower()
            
            if answer != 'y':
                print("\n❌ 迁移已取消")
                conn.close()
                return
            
            # 清空现有数据
            cursor.execute("DELETE FROM sessions")
            cursor.execute("DELETE FROM users")
            conn.commit()
            print("✅ 已清空现有用户数据\n")
        
        # 插入新用户
        success_count = 0
        fail_count = 0
        
        for user in USERS_TO_MIGRATE:
            try:
                salt = generate_salt()
                password_hash = hash_password(DEFAULT_PASSWORD, salt)
                
                cursor.execute("""
                    INSERT INTO users 
                    (username, password_hash, salt, role, permissions)
                    VALUES (?, ?, ?, ?, ?)
                """, (
                    user['username'],
                    password_hash,
                    salt,
                    user['role'],
                    str(user['permissions'])
                ))
                
                if cursor.rowcount > 0:
                    print(f"  ✅ {user['username']:12s} | 角色: {user['role']:10s} | 密码: {DEFAULT_PASSWORD}")
                    success_count += 1
                else:
                    print(f"  ❌ {user['username']:12s} | 插入失败")
                    fail_count += 1
                    
            except sqlite3.IntegrityError as e:
                if 'UNIQUE' in str(e):
                    print(f"  ⚠️ {user['username']:12s} | 已存在，跳过")
                else:
                    print(f"  ❌ {user['username']:12s} | 错误: {e}")
                    fail_count += 1
            except Exception as e:
                print(f"  ❌ {user['username']:12s} | 错误: {e}")
                fail_count += 1
        
        conn.commit()
        
        # 输出结果
        print()
        print("=" * 60)
        print(f"📊 迁移结果:")
        print(f"  ✅ 成功: {success_count} 个用户")
        print(f"  ❌ 失败: {fail_count} 个用户")
        print()
        print(f"📝 登录信息:")
        print(f"  用户名: {'admin':12s} | 密码: {DEFAULT_PASSWORD:10s} | 角色: 管理员")
        print(f"  用户名: {'operator':12s} | 密码: {DEFAULT_PASSWORD:10s} | 角色: 操作员")
        print(f"  用户名: {'viewer':12s} | 密码: {DEFAULT_PASSWORD:10s} | 角色: 访客")
        print()
        print("⚠️ 请立即修改默认密码！")
        print("=" * 60)
        
        conn.close()
        
    except Exception as e:
        print(f"\n❌ 迁移失败: {e}")
        import traceback
        traceback.print_exc()
        sys.exit(1)


if __name__ == '__main__':
    # 获取数据库路径参数
    if len(sys.argv) > 1:
        db_path = sys.argv[1]
    else:
        # 默认路径
        db_path = os.path.join(
            os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
            'data',
            'jobs.db'
        )
    
    migrate_users(db_path)

File: scripts/migrate-users.ts
/**
 * 用户数据迁移脚本
 *
 * 将现有的环境变量用户迁移到数据库
 * 默认密码: root
 * 角色: admin / operator / viewer
 */

import Database from 'better-sqlite3';
import { createHash, randomBytes } from 'crypto';
import path from 'path';

// ==================== 配置 ====================

const DB_PATH = process.argv[2] || path.join(process.cwd(), 'data', 'jobs.db');
const DEFAULT_PASSWORD = 'root'; // 用户指定的默认密码

// 要迁移的用户列表 (从 .env 迁移)
const USERS_TO_MIGRATE = [
  {
    username: 'admin',
    role: 'admin',
    permissions: [
      'view_jobs', 'view_stats', 'manage_crawler', 'view_system',
      'use_recommendations', 'manage_users',
      'jobs:read', 'jobs:write', 'crawler:read', 'crawler:write',
      'system:read', 'system:write', 'users:read', 'users:write',
      'recommendations:read', 'recommendations:write',
      'match:read', 'match:write',
    ]
  },
  {
    username: 'operator',
    role: 'operator',
    permissions: [
      'view_jobs', 'view_stats', 'manage_crawler', 'view_system',
      'use_recommendations',
      'jobs:read', 'jobs:write', 'crawler:read', 'crawler:write',
      'system:read', 'recommendations:read', 'recommendations:write',
      'match:read', 'match:write',
    ]
  },
  {
    username: 'viewer',
    role: 'viewer',
    permissions: [
      'view_jobs', 'view_stats', 'view_system', 'use_recommendations',
      'jobs:read', 'crawler:read', 'system:read',
      'recommendations:read', 'match:read',
    ]
  }
];

// ==================== 工具函数 ====================

function generateSalt(length: number = 32): string {
  return randomBytes(length).toString('hex');
}

function hashPassword(password: string, salt: string): string {
  return createHash('sha256').update(`${password}${salt}`).digest('hex');
}

// ==================== 主逻辑 ====================

async function migrateUsers() {
  console.log('=' .repeat(60));
  console.log('🚀 开始迁移用户到数据库...');
  console.log(`📁 数据库路径: ${DB_PATH}`);
  console.log(`🔑 默认密码: ${DEFAULT_PASSWORD}`);
  console.log('');

  let db: Database.Database | undefined;

  try {
    // 连接数据库
    db = new Database(DB_PATH);
    
    // 启用 WAL 模式
    db.pragma('journal_mode = WAL');

    console.log('✅ 数据库连接成功\n');

    // 创建认证表 (如果不存在)
    db.exec(`
      CREATE TABLE IF NOT EXISTS users (
        id INTEGER PRIMARY KEY AUTOINCREMENT,
        username TEXT UNIQUE NOT NULL,
        password_hash TEXT NOT NULL,
        salt TEXT NOT NULL,
        role TEXT NOT NULL DEFAULT 'viewer',
        permissions TEXT DEFAULT '[]',
        is_active INTEGER DEFAULT 1,
        failed_login_count INTEGER DEFAULT 0,
        locked_until TEXT NULL,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        updated_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        last_login_at TIMESTAMP NULL
      )
    `);

    db.exec(`
      CREATE TABLE IF NOT EXISTS sessions (
        id TEXT PRIMARY KEY,
        user_id INTEGER NOT NULL,
        token TEXT UNIQUE NOT NULL,
        ip_address TEXT,
        user_agent TEXT,
        expires_at TIMESTAMP NOT NULL,
        created_at TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
        FOREIGN KEY (user_id) REFERENCES users(id) ON DELETE CASCADE
      )
    `);

    console.log('✅ 认证表已就绪\n');

    // 检查是否已有用户
    const existingUsers = db.prepare('SELECT COUNT(*) as count FROM users').get() as any;
    
    if (existingUsers.count > 0) {
      console.log(`⚠️ 数据库中已有 ${existingUsers.count} 个用户`);
      
      const readline = require('readline');
      const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
      });

      const answer = await new Promise<string>((resolve) => {
        rl.question('是否继续覆盖？(y/N): ', resolve);
      });
      rl.close();

      if (answer.toLowerCase() !== 'y') {
        console.log('\n❌ 迁移已取消');
        db.close();
        return;
      }

      // 清空现有用户和会话
      db.exec('DELETE FROM sessions');
      db.exec('DELETE FROM users');
      console.log('✅ 已清空现有用户数据\n');
    }

    // 插入新用户
    let successCount = 0;
    let failCount = 0;

    for (const user of USERS_TO_MIGRATE) {
      try {
        const salt = generateSalt();
        const passwordHash = hashPassword(DEFAULT_PASSWORD, salt);

        const result = db.prepare(`
          INSERT INTO users (username, password_hash, salt, role, permissions)
          VALUES (?, ?, ?, ?, ?)
        `).run(user.username, passwordHash, salt, user.role, JSON.stringify(user.permissions));

        if (result.changes > 0) {
          console.log(`  ✅ ${user.username.padEnd(12)} | 角色: ${user.role.padEnd(10)} | 密码: ${DEFAULT_PASSWORD}`);
          successCount++;
        } else {
          console.log(`  ❌ ${user.username.padEnd(12)} | 插入失败`);
          failCount++;
        }

      } catch (error: any) {
        if (error.message?.includes('UNIQUE constraint')) {
          console.log(`  ⚠️ ${user.username.padEnd(12)} | 已存在，跳过`);
        } else {
          console.error(`  ❌ ${user.username.padEnd(12)} | 错误: ${error.message}`);
          failCount++;
        }
      }
    }

    console.log('');
    console.log('=' .repeat(60));
    console.log(`📊 迁移结果:`);
    console.log(`  ✅ 成功: ${successCount} 个用户`);
    console.log(`  ❌ 失败: ${failCount} 个用户`);
    console.log('');
    console.log(`📝 登录信息:`);
    console.log(`  用户名: admin     | 密码: ${DEFAULT_PASSWORD} | 角色: 管理员`);
    console.log(`  用户名: operator  | 密码: ${DEFAULT_PASSWORD} | 角色: 操作员`);
    console.log(`  用户名: viewer    | 密码: ${DEFAULT_PASSWORD} | 角色:访客`);
    console.log('');
    console.log('⚠️ 请立即修改默认密码！');
    console.log('=' .repeat(60));

    db.close();

  } catch (error) {
    console.error('❌ 迁移失败:', error);
    if (db) db.close();
    process.exit(1);
  }
}

// 执行迁移
migrateUsers();
